README.md
# ALB Manager Serverless Plugin

[![CircleCI](https://circleci.com/gh/schoology/serverless-alb-manager.svg?style=shield)](https://circleci.com/gh/schoology/serverless-alb-manager)
[![npm version](https://badge.fury.io/js/serverless-alb-manager.svg)](https://badge.fury.io/js/serverless)
[![license](https://img.shields.io/npm/l/serverless-alb-manager.svg)](https://www.npmjs.com/package/serverless-alb-manager)

This plugin adds an HTTP ALB to your serverless project instead of an API Gateway for HTTP functions.

## About Schoology

<p align="center"><img src="https://app.schoology.com/sites/all/themes/schoology_theme/images_front/schoology_thumbnail.png" width="120" alt="Schoology Logo" /></p>

Schoology, one of the fastest growing education technology companies, brings together the best K-12 learning management 
system with assessment management to improve student performance, foster collaboration, and personalize learning.
Millions of students, faculty and administrators from over 60,000 K-12 schools worldwide use Schoology to advance what is possible in education.

We help solve the most important challenges in education in an environment that’s fun, collaborative and innovative.
We are looking for smart, creative individuals who are passionate about education and aren't afraid to show it.

Learn more at https://www.schoology.com/careers.

## Requirements
Tested with:
* Node.js >= `v8.10`
* Serverless Framework >= `v1.45`

## Installation

1. Using yarn:
    ```
    yarn add --dev serverless-alb-manager
    ```

2. Add the plugin to your `serverless.yml` file:
    ```yaml
    plugins:
      - serverless-alb-manager
    ```

3. Add the plugin configuration to your `serverless.yml` file, for example:
    ```yml
      serverless-alb-manager:
        vpcId: vpc-1234
        subnetIds:
          - subnet-2345
          - subnet-3456
        tags:
          foo: fooValue
          bar: barValue
        certificateArn: "arn:aws:acm:us-east-1:4567:certificate/abcd-5678"
        domainName: alb.example.com
    ```
    
    |Parameter Name|Required|Description|
    |---|---|---|
    |`vpcId`|:heavy_check_mark:|The VPC in which the ALB should be created.|
    |`subnetIds`|:heavy_check_mark:|An array of subnet IDs to associate with the ALB. These subnets should belong to the configured `vpcId`.|
    |`tags`|:heavy_check_mark:|A map of tags for the ALB.|
    |`certificateArn`|:heavy_check_mark:|The AWS ACM certificate ARN to use for the HTTP listener.|
    |`domainName`|:heavy_check_mark:|The domain name that will be used to access the ALB. This should be valid for the given `certificateArn`.|

4. Add alb event listeners to your serverless functions. The `listenerArn` will
automatically be populated with the ALB listener created by this plugin. For example:

    ```yaml
    functions:
      foo:
        handler: src/handler.foo
        events:
          - alb:
              # libraryArn is automatically populated
              priority: 1
              conditions:
                path: '/foo'
                method: GET
       bar:
        handler: src/handler.bar
        events:
          - alb:
              # libraryArn is automatically populated
              priority: 2
              conditions:
                path: '/bar'
                method: GET
   
    ```

## Resource Created By This Plugin

|Type|Description|
|---|---|
|**Security Group**|<ul><li>ingress `443` from `0.0.0.0/0`</li><li>egress `*` to `0.0.0.0/0`</li><li>configurable VPC</li></ul>|
|**Application Load Balancer**|<ul><li>ipv4</li><li>configurable subnet</li><li>configurable tags</li></ul>|
|**ALB Listener**|<ul><li>default action `HTTP 403` with JSON body `{ "error": "Forbidden" }`</li><li>configurable ACM certificate ARN</li></ul>|
|**Route53 DNS Record**|<ul><li>Alias pointing to ALB</li><li>Configurable domain name</li></ul>|


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.3](https://github.com/schoology/serverless-alb-manager/compare/v1.0.2...v1.0.3) (2020-01-27)

### 1.0.2 (2019-11-04)


src/index.spec.ts
import * as fs from 'fs';
import * as Serverless from 'serverless';
import Service = require('serverless/classes/Service');
import * as TypeMoq from 'typemoq';

import { ServerlessPluginAlbManager as ServerlessPluginAlbManagerType } from './index';
import { AlbManagerOptions } from './types';

// Import class separately from types since the plugin uses CommonJS export
const ServerlessPluginAlbManager = require('./index'); // tslint:disable-line
const ServiceClass = require('serverless/lib/classes/Service'); // tslint:disable-line

/**
 * Mock a `custom.serverless-alb-manager` config from the Service mock
 */
const setConfig = (service: Service, config: object): void => {
  service.update({ custom: { ['serverless-alb-manager']: config } });
};

describe('ServerlessPluginAlbManager', () => {
  let serverlessMock: TypeMoq.IMock<Serverless>;
  let service: Service;
  let plugin: ServerlessPluginAlbManagerType;
  let providerObject: Service['provider'];
  let validConfig: AlbManagerOptions;

  beforeEach(() => {
    providerObject = {
      compiledCloudFormationTemplate: {
        Outputs: {},
        Resources: {},
      },
      name: 'dummy-provider-name',
      stage: 'dummy-provider-stage',
      region: 'dummy-provider-region',
      versionFunctions: true,
      naming: {
        getStackName: () => 'dummy-stack-name',
      },
    };

    validConfig = {
      vpcId: 'dummy-vpc-id',
      subnetIds: ['dummy-subnet-1', 'dummy-subnet-2'],
      certificateArn: 'dummy-certificate-arn',
      domainName: 'dummy.domain.example.com',
      tags: {
        dummyTag1: 'dummy-1',
        dummyTag2: 'dummy-2',
      },
    };

    serverlessMock = TypeMoq.Mock.ofType<Serverless>();

    service = new ServiceClass(serverlessMock.object, {
      provider: providerObject,
    });

    serverlessMock.setup(x => x.service).returns(() => service);
    serverlessMock.setup(x => x.getProvider(TypeMoq.It.isAnyString())).returns(() => providerObject);

    plugin = new ServerlessPluginAlbManager(serverlessMock.object);
  });

  describe('package:setupProviderConfiguration hook', () => {
    describe('with all expected config options', () => {
      beforeEach(() => {
        setConfig(service, validConfig);
      });

      it('generates a valid cloudformation template', async () => {
        await plugin.hooks['package:setupProviderConfiguration']();
        expect(service.provider.compiledCloudFormationTemplate).toMatchSnapshot();

        // Write out a JSON that will be validated during CI
        fs.writeFileSync(
          'template-snapshot.json',
          // tslint:disable-next-line:no-magic-numbers
          JSON.stringify(service.provider.compiledCloudFormationTemplate, undefined, 2),
        );
      });
    });

    describe('without tags', () => {
      beforeEach(() => {
        setConfig(service, {
          ...validConfig,
          tags: undefined,
        });
      });

      it('does not set any tags', async () => {
        await plugin.hooks['package:setupProviderConfiguration']();
        expect(service.provider.compiledCloudFormationTemplate.Resources.Alb.Properties.Tags).toBeUndefined();
      });
    });

    describe('with comma-separated subnet IDs', () => {
      beforeEach(() => {
        setConfig(service, {
          ...validConfig,
          subnetIds: 'dummy-subnet-3, dummy-subnet-4',
        });
      });

      it('splits the subnets into an array', async () => {
        await plugin.hooks['package:setupProviderConfiguration']();
        expect(service.provider.compiledCloudFormationTemplate.Resources.Alb.Properties.Subnets).toMatchObject([
          'dummy-subnet-3',
          'dummy-subnet-4',
        ]);
      });
    });

    describe('with no config options', () => {
      it('throws an error', () => {
        return expect(plugin.hooks['package:setupProviderConfiguration']()).rejects.toThrowError(
          'serverless-plugin-alb-manager: Please define ALB options in serverless.yml:custom.alb. See README.md for details.',
        );
      });
    });

    describe('with invalid config options', () => {
      beforeEach(() => {
        setConfig(service, {
          foo: 'bar',
        });
      });

      it('throws an error', () => {
        return expect(plugin.hooks['package:setupProviderConfiguration']()).rejects.toThrowError(
          'serverless-plugin-alb-manager: Invalid configuration. ValidationError: ' +
            '"vpcId" is required. ' +
            '"subnetIds" is required. ' +
            '"certificateArn" is required. ' +
            '"domainName" is required. ' +
            '"foo" is not allowed',
        );
      });
    });
  });

  describe('package:compileFunctions hook', () => {
    describe('with alb events that do not have listenerArn defined', () => {
      beforeEach(() => {
        service.update({
          functions: {
            dummyFunc1: {
              events: [{ alb: { conditions: { path: '/dummy11' } } }, { alb: { conditions: { path: '/dummy12' } } }],
            },
            dummyFunc2: {
              events: [{ alb: { conditions: { path: '/dummy21' } } }, { alb: { conditions: { path: '/dummy22' } } }],
            },
          },
        });
        service.setFunctionNames({});
      });

      it('sets the event listenerArns', async () => {
        await plugin.hooks['package:compileFunctions']();
        const expectedListener = {
          Ref: 'AlbHttpListener',
        };
        expect(service.getAllEventsInFunction('dummyFunc1')[0].alb.listenerArn).toMatchObject(expectedListener);
        expect(service.getAllEventsInFunction('dummyFunc1')[1].alb.listenerArn).toMatchObject(expectedListener);
        expect(service.getAllEventsInFunction('dummyFunc2')[0].alb.listenerArn).toMatchObject(expectedListener);
        expect(service.getAllEventsInFunction('dummyFunc2')[1].alb.listenerArn).toMatchObject(expectedListener);
      });
    });

    describe('with alb events that do have listenerArn defined', () => {
      beforeEach(() => {
        service.update({
          functions: {
            dummyFunc1: {
              events: [
                {
                  alb: {
                    conditions: { path: '/dummy11' },
                    listenerArn: 'dummyListenerArn',
                  },
                },
              ],
            },
          },
        });
        service.setFunctionNames({});
      });

      it('sets the event listenerArns', async () => {
        await plugin.hooks['package:compileFunctions']();
        expect(service.getAllEventsInFunction('dummyFunc1')[0].alb.listenerArn).toBe('dummyListenerArn');
      });
    });

    describe('with non-alb events', () => {
      const nonAlbEvents = [{ http: 'GET /dummyFunc1' }];
      beforeEach(() => {
        service.update({
          functions: {
            dummyFunc1: { events: nonAlbEvents },
          },
        });
        service.setFunctionNames({});
      });

      it('does not modify the events', async () => {
        await plugin.hooks['package:compileFunctions']();
        expect(service.getAllEventsInFunction('dummyFunc1')).toMatchObject(nonAlbEvents);
      });
    });
  });
});


types/serverless/index.d.ts
// serverless type declaration fixes - to be removed once
// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38315 is merged and released

// tslint:disable:no-any max-classes-per-file

declare module 'serverless/classes/Plugin' {
  const serverlessOriginal = require('serverless');
  declare interface Plugin {
    hooks: {
      [event: string]: () => Promise<any>;
    };
  }

  declare var Plugin: {
    prototype: Plugin;
    new (serverless: Serverless, options: Serverless.Options): Plugin;
  };
  export = Plugin;
}

declare module 'serverless/classes/Service' {
  declare class Service {
    public custom: Service.Custom;

    public provider: {
      compiledCloudFormationTemplate: {
        Resources: {
          [key: string]: any;
        };
        Outputs?: {
          [key: string]: any;
        };
      };

      name: string;
      stage: string;
      region: string;
      runtime?: string;
      timeout?: number;
      versionFunctions: boolean;

      naming: Record<string, () => string>;
    };
    constructor(serverless: Serverless, data: {});

    public load(rawOptions: {}): Promise<any>;
    public setFunctionNames(rawOptions: {}): void;

    public getServiceName(): string;
    public getAllFunctions(): string[];
    public getAllFunctionsNames(): string[];
    public getFunction(functionName: string): Serverless.FunctionDefinition;
    public getEventInFunction(eventName: string, functionName: string): Serverless.Event;
    public getAllEventsInFunction(functionName: string): Serverless.Event[];

    public mergeResourceArrays(): void;
    public validate(): Service;

    public update(data: {}): {};
  }

  export = Service;
}

declare module 'serverless/plugins/aws/provider/awsProvider' {
  declare class Aws {
    public naming: Record<string, () => string>;
    constructor(serverless: Serverless, options: Serverless.Options);
    public getProviderName(): string;
    public getRegion(): string;
    public getServerlessDeploymentBucketName(): string;
    public getStage(): string;
  }

  export = Aws;
}

declare module 'serverless' {
  const serverlessOriginal = require('serverless');

  declare namespace Serverless {
    export declare interface Event {
      alb?: {
        listenerArn?: string | object;
      };
    }
  }

  declare class Serverless extends serverlessOriginal {}

  export = Serverless;
}


src/index.ts
import * as Joi from '@hapi/joi';
import * as Serverless from 'serverless';
import Plugin = require('serverless/classes/Plugin');
import AwsProvider = require('serverless/plugins/aws/provider/awsProvider');

import { AlbManagerOptions } from './types';

export class ServerlessPluginAlbManager implements Plugin {
  public hooks: Record<string, () => Promise<any>>; // tslint:disable-line:no-any
  private provider: AwsProvider;

  constructor(private serverless: Serverless) {
    this.provider = serverless.getProvider('aws') as AwsProvider;

    this.hooks = {
      'package:setupProviderConfiguration': this.compileAlb.bind(this),
      'package:compileFunctions': this.setEventListenerArn.bind(this),
    };
  }

  public async compileAlb(): Promise<void> {
    const resources = this.serverless.service.provider.compiledCloudFormationTemplate.Resources;
    const options = await this.getConfigOptions();

    // Construct CF resources
    this.addSecurityGroup(resources, options.vpcId);
    this.addLoadBalancer(resources, options.subnetIds, options.tags);
    this.addHttpListener(resources, options.certificateArn);
    this.addDnsRecord(resources, options.domainName);
  }

  public setEventListenerArn(): Promise<void> {
    const functions = this.serverless.service.getAllFunctions();
    functions.forEach((functionName: string) => {
      this.serverless.service
        .getAllEventsInFunction(functionName)
        .forEach((event: Serverless.Event, eventNum: number) => {
          // If this event is an alb event but listenerArn is not defined,
          // populate the listenerArn with the ALB Listener we created
          if (event.alb && !event.alb.listenerArn) {
            this.serverless.service.update({
              functions: {
                [functionName]: {
                  events: {
                    [eventNum]: {
                      alb: {
                        listenerArn: { Ref: 'AlbHttpListener' },
                      },
                    },
                  },
                },
              },
            });
          }
        });
    });
    return Promise.resolve();
  }

  private async getConfigOptions(): Promise<AlbManagerOptions> {
    if (!this.serverless.service.custom || !this.serverless.service.custom['serverless-alb-manager']) {
      throw new Error(
        'serverless-plugin-alb-manager: Please define ALB options in serverless.yml:custom.alb. See README.md for details.',
      );
    }

    const schema = Joi.object({
      vpcId: Joi.string().required(),
      subnetIds: Joi.alternatives()
        // Either a comma separated string, or an array of strings
        .try(
          Joi.string(),
          Joi.array()
            .items(Joi.string())
            .min(1),
        )
        .required(),
      tags: Joi.object(),
      certificateArn: Joi.string().required(),
      domainName: Joi.string()
        .domain()
        .required(),
    });

    const { value, error } = schema.validate(this.serverless.service.custom['serverless-alb-manager'], {
      abortEarly: false,
    });
    if (error) {
      throw new Error('serverless-plugin-alb-manager: Invalid configuration. ' + error.toString());
    }

    // Split comma-separated values
    if (typeof value.subnetIds === 'string') {
      value.subnetIds = value.subnetIds.split(',').map((x: string) => x.trim());
    }

    return value;
  }

  private addSecurityGroup(resources: object, vpcId: AlbManagerOptions['vpcId']): void {
    Object.assign(resources, {
      AlbSecurityGroup: {
        Type: 'AWS::EC2::SecurityGroup',
        Properties: {
          GroupName: `${this.provider.naming.getStackName()}-https`,
          GroupDescription: `HTTPS for ${this.provider.naming.getStackName()}`,
          VpcId: vpcId,
          SecurityGroupIngress: [
            {
              IpProtocol: 'tcp',
              FromPort: 443,
              ToPort: 443,
              CidrIp: '0.0.0.0/0',
            },
          ],
        },
      },
    });
  }

  private addLoadBalancer(
    resources: object,
    subnetIds: AlbManagerOptions['subnetIds'],
    tags: AlbManagerOptions['tags'],
  ): void {
    const cfTags = tags ? Object.keys(tags).map(key => ({ Key: key, Value: tags[key] })) : undefined;

    Object.assign(resources, {
      Alb: {
        Type: 'AWS::ElasticLoadBalancingV2::LoadBalancer',
        Properties: {
          Type: 'application',
          Name: `${this.provider.naming.getStackName()}-alb`,
          IpAddressType: 'ipv4',
          Scheme: 'internet-facing',
          SecurityGroups: [
            {
              Ref: 'AlbSecurityGroup',
            },
          ],
          Subnets: subnetIds,
          Tags: cfTags,
        },
      },
    });
  }

  private addHttpListener(resources: object, certificateArn: AlbManagerOptions['certificateArn']): void {
    Object.assign(resources, {
      AlbHttpListener: {
        Type: 'AWS::ElasticLoadBalancingV2::Listener',
        Properties: {
          LoadBalancerArn: {
            Ref: 'Alb',
          },
          Port: 443,
          Protocol: 'HTTPS',
          Certificates: [
            {
              CertificateArn: certificateArn,
            },
          ],
          DefaultActions: [
            {
              Type: 'fixed-response',
              Order: 1,
              FixedResponseConfig: {
                StatusCode: '403',
                ContentType: 'application/json',
                MessageBody: '{ "error": "Forbidden" }',
              },
            },
          ],
        },
      },
    });
  }

  private addDnsRecord(resources: object, domainName: AlbManagerOptions['domainName']): void {
    Object.assign(resources, {
      AlbDnsRecord: {
        Type: 'AWS::Route53::RecordSet',
        Properties: {
          HostedZoneName: this.getRootDomain(domainName) + '.',
          Name: domainName,
          Type: 'A',
          AliasTarget: {
            DNSName: {
              'Fn::GetAtt': ['Alb', 'DNSName'],
            },
            HostedZoneId: {
              'Fn::GetAtt': ['Alb', 'CanonicalHostedZoneID'],
            },
          },
        },
      },
    });
  }

  private getRootDomain(domainName: string) {
    const domainSplit = domainName.split('.').reverse();
    return domainSplit[1] + '.' + domainSplit[0];
  }
}

module.exports = ServerlessPluginAlbManager;
